### ISSUE TITLE
When deleting a content type add option to delete content

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#2461]
A checkbox on the confirm form would do.

**Comment by docwilmot** [comment]
More info:
- currently youre allowed to delete a content type with nodes, with a warning

![cntnt type delete](https://cloud.githubusercontent.com/assets/1042283/22036474/b43d2008-dcc1-11e6-8606-5a2f9fa2f8e1.PNG)


- if you delete a content type the nodes remaining have no type specified:

![foo deleted](https://cloud.githubusercontent.com/assets/1042283/22036446/94680306-dcc1-11e6-9a58-c47d2392ad68.PNG)


- attempt to edit the ghost nodes gives a blank edit page and attempt to view gives a server error.

This long discussion on Drupal led to the decision to actually block deletion of entity types with remaining content (link to a specific suggestion, same as this issue): https://www.drupal.org/node/232327#comment-3274338

We probably need to repurpose this issue to what to do about deleting content types with remaining content, but for now I'll leave the title as this subset of that discussion.

**Comment by Dinornis** [comment]
Great idea.

**Comment by klonos** [comment]
It would be really great if we offered the option to export/save the data before deletion. Sorry, I come from a data-security background and I am always (over)conscious when destructive operations are being performed over data: #2498

**Comment by docwilmot** [comment]
I've been a bit conflicted over this issue actually. The Drupal way of preventing deletion of the content type if there are remaining nodes makes sense; it would prevent data loss better than the option presented here.

#2498 is a good idea long term, but still doesnt help this issue. Even if you trash/unpublish the nodes, you cant delete the content type as you would still be left with ghost nodes. Except if we also find a way to trash content types and fields instead of deleting them.

Similarly, exporting will only work if we can also export the content types and fields too.

**Comment by jenlampton** [comment]
I'd love to see a UI Improvement that prevents deleting a content type until all the nodes have been deleted, but since we are only 1 day from feature freeze, it's unlikely the new feature "option to delete content" will make it into this release.

I'm going to bump the milestone on this feature to 1.8, but if someone remains motivated enough to fix the content deletion bug on 1.7 without the new feature, please create a new issue and add the milestone.

**Comment by jenlampton** [comment]
We're two weeks away from code freeze for 1.8, and with no code here yet to review or revise it's not likely this feature will get done in time. Removing the milestone.

**Comment by klonos** [comment]
I have reread the comments in this issue, and here are my thoughts:

> ...Except if we also find a way to trash content types and fields instead of deleting them.

I like this idea. Here's what we could do:

- Offer a way to disable/enable content types (same as we do for views and layouts).
- Introduce a "Content type status" views filter and sort.
- When a content type is disabled, use the before-mentioned filter to hide the respective pieces of content from the content listing page in `/admin/content` (and from other views).
- When listing content types in `/admin/structure/types`, administrators should be able to view both enable/disabled types.
- When listing content types in `/node/add`, disabled content types would be hidden.

...the above is not a prerequisite for moving this ticket forward, but would limit the use cases where a content type would need to be deleted to only when people are actually determined to get rid of the content type + all the respective content. So then we'd then only need to decide to either:

- Add a checkbox and a big, red warning in `/admin/structure/types/manage/[content-type]/delete` to confirm deleting all content before deleting a content type (if content of that type already exists)

...or:

- Disallow deleting a content type while content of that type still exists.

**Comment by jenlampton** [comment]
@klonos that sounds like a lot of work. Sanity check: is this an 80% use-case feature? I'd rather see the fix that prevents types from being deleted while there is still content, and this rest of this be handled in contrib.

**Comment by klonos** [comment]
OK, I took a stab at a PR...

- If content of that type exists:

    <img width="600" alt="screen shot 2018-12-12 at 6 14 17 am" src="https://user-images.githubusercontent.com/2423362/49846723-52929680-fdd5-11e8-9275-66a953234794.png">

- If no content of that type exists (screenshot is from my local, which includes #3422):

    <img width="626" alt="screen shot 2018-12-12 at 6 14 08 am" src="https://user-images.githubusercontent.com/2423362/49846726-5a523b00-fdd5-11e8-93b7-06540e929b47.png">

**Comment by klonos** [comment]
...here's current D8 for reference btw:

<img width="684" alt="screen shot 2018-12-12 at 11 15 41 am" src="https://user-images.githubusercontent.com/2423362/49859137-77036880-fdff-11e8-9411-971094bf0cf9.png">

<img width="877" alt="screen shot 2018-12-12 at 11 16 12 am" src="https://user-images.githubusercontent.com/2423362/49859143-78349580-fdff-11e8-8b92-a2631f63c1a0.png">

**Comment by olafgrabienski** [comment]
> To remove this content type, you first need to *remove all content of that type*.

Great, that there's a link to the `admin/content` page on the "remove" part of the sentence. I'd suggest to filter the page to only display items of the respective content type (as in the result of `admin/content/node?status=All&type=MY-TYPE`).

**Comment by klonos** [comment]
...after popular demand (😜), an alternative PR that allows deleting content of the type to be deleted (with confirmation via a checkbox):

Screenshots (include fixes for #3422 and #3425):

- If no content of that type exists:

    <img width="680" alt="screen shot 2018-12-12 at 4 40 58 pm" src="https://user-images.githubusercontent.com/2423362/49876954-10953f00-fe2d-11e8-852c-8a6e9e649c3e.png">

- If content of that type exists (delete button is enabled only after the checkbox is ticked):

    <img width="683" alt="screen shot 2018-12-12 at 4 39 11 pm" src="https://user-images.githubusercontent.com/2423362/49876930-07a46d80-fe2d-11e8-998a-c9729a098f19.png">

    <img width="685" alt="screen shot 2018-12-12 at 4 39 24 pm" src="https://user-images.githubusercontent.com/2423362/49876940-0bd08b00-fe2d-11e8-994f-055c58d7d844.png">

...after confirming and deleting:

    <img width="484" alt="screen shot 2018-12-12 at 4 44 56 pm" src="https://user-images.githubusercontent.com/2423362/49877100-53571700-fe2d-11e8-8093-ef290f730d5f.png">

**Comment by docwilmot** [comment]
i like this best

**Comment by docwilmot** [comment]
I would use "delete" instead of "remove" though, just to be perfectly clear.

**Comment by klonos** [comment]
> ...I'd suggest to filter the page to only display items of the respective content type...

Great suggestion @olafgrabienski 👍 ...I've updated the PR to include that.

> I would use "delete" instead of "remove" though, just to be perfectly clear.

Yep, I missed that @docwilmot. Fixed in both PRs 😉

**Comment by olafgrabienski** [comment]
I've tested https://github.com/backdrop/backdrop/pull/2420 in the sandbox. After clicking the button `Delete all content and the content type` I got this message:

> - Deleted 2 pieces of content of type *Article*.
> - The content type *Article* has been deleted.

As a result, the articles have been deleted but the content type not.

**Comment by jenlampton** [comment]
I'm not able to get this PR working: https://github.com/backdrop/backdrop/pull/2420

**Comment by jenlampton** [comment]
Removing the 1.12 milestone, but adding the tag for review later.

**Comment by jenlampton** [comment]
removing `[D8 feature parity]` from title, adding into project.

**Comment by klonos** [comment]
I started working on this again...

@olafgrabienski you are right, "content item(s) of type ..." sounds better than "piece(s) of content of type..." in this instance. Update the PR accordingly.

> I'm not able to get this PR working: backdrop/backdrop#2420

I think that it needed to be rebased @jenlampton. Can you please give it another go now?

**Comment by klonos** [comment]
...leaving the NW label, as I plan to implement the following:

- link the `There is x content item(s) of type test content type on your site.` bit to the content listing, filter by the content type to be deleted.
- write a new test for testing the functionality of deleting a content type w/o any content vs deleting one that has existing content.

**Comment by alanmels** [comment]
I've just tested this on http://2420.backdrop.backdrop.qa.backdropcms.org/admin/structure/types and unfortunately it still does not delete the content type itself. The website's log shows: 

```
Error
Argument 2 passed to t() must be an array, integer given, called in /home/qa/github/backdrop/backdrop/pr/2420/core/modules/dblog/dblog.admin.inc on line 310 and defined
```

Is the version on http://2420.backdrop.backdrop.qa.backdropcms.org the most up-to-date or should I download the PR to localhost to test?

**Comment by jenlampton** [comment]
@alanmels The current PR (at http://2420.backdrop.backdrop.qa.backdropcms.org/) needs to be rebased. I'm not sure if that will fix the problem you are seeing, but the results will be the same if you download the PR to test locally.

**Comment by alanmels** [comment]
@jenlampton, thanks! Then I'll try later.

**Comment by philsward** [comment]
Any progress on this?

I don't think this issue is a show-stopper but it does create the potential for a lot of bad UX for users. Most content types are deleted early during restructuring, however if there is a content type that is deleted and other nodes link to pages of the deleted content type, you're presented at most with a bunch of PHP errors instead of a proper 404.

At the very least you get:
```
site.com
**Error**
Call to a member function fetchCol() on null
```

This isn't very welcoming if a normal user navigates to this page.

One thought as a temporary workaround until this can be fixed proper is to present an error that is more friendly, wraps the error in the sites theme and says "The page you are looking for is pending deletion.". Put a simple check in the code that says "if content_type = null, show pending deletion message". (I'm not a coder, but hopefully that makes sense)

The other temporary workaround thought is to move all of the content to a reserved (hidden) content type and unpublish them. This should give an access denied error I believe which is better than the current error UX.

---

As far as what @klonos has suggested in https://github.com/backdrop/backdrop-issues/issues/2461#issuecomment-446612198, I personally love that implementation. It gives an actionable confirmation to the user and it is clear on what needs to happen.

I will say, it might be really cool if it could traverse through all the other modules that touch the content type. I don't think it needs to hard stop, just show the number of entities that are being deleted and then show something like:

In addition, deleting this content type will affect the following:
* "xyz view"
* "abc view"
* "yzf rule"

Maybe add a checkbox to confirm?

In addition, deleting this content type will affect the following:
* "xyz view"
* "abc view"
* "yzf rule"
[  ] I understand

**Comment by stpaultim** [comment]
Adding this to my watch list. Seems like a problem that needs a solution and it seems like we've gotten close to solving it in the past. 

As far as I can tell, the next step is for someone to look at the most recent PR and see if they can determine why it's not working as expected.

**Comment by indigoxela** [comment]
@klonos I've left two comments at your PR.

**Comment by philsward** [comment]
One thing I've wondered regarding this particular issue, how are entities in general handled in this regard?

Here we're talking about content types and nodes, what about paragraphs, or field collections or commerce products as examples? Should the same or a similar approach be taken with all entities or am I over thinking it?

Paragraphs probably won't really be an issue because there isn't an entity type container, but products with commerce (for example) will because the products are housed within an entity type. I say that... The more I think about it, commerce may be a bad example because a product is locked from deletion once it's been added to an order. I have very little experience with field collection as Paragraphs became my primary go-to.

Mainly just bringing up the idea that this may need to extend beyond nodes or at least have the ability for other modules to tie into its functionality.

**Comment by indigoxela** [comment]
> Mainly just bringing up the idea that this may need to extend beyond nodes

If your concern's about things _attached_ to nodes - that's a job for hook_node_delete/hook_entity_delete which (contrib) modules are supposed to implement.

Custom file types might deserve some attention. But I suppose, that could/should be a follow-up issue.

**Comment by philsward** [comment]
> If your concern's about things _attached_ to nodes ...

No, the concern is with contribs that create their own entities inside their own entity types.

ECK should be an excellent example though I don't know if it's been ported yet.

**Comment by klonos** [comment]
Right. Thanks for reviving this @philsward 👍 ...I should get back to it then.

**Comment by klonos** [comment]
...also thanks @indigoxela for the PR review, and @jenlampton, @alanmels, and @stpaultim for the feedback and ideas 🙏 

I also find this a handy UX improvement, and in the meantime one of the 2 issues that were "supplementing" this has been fixed/merged (talking about #3422 and #3425). I'll get back to working on this over the next few days (the weekend most likely). Thanks again all.

**Comment by stpaultim** [comment]
@klonos - is there room for this on your to-do list?

**Comment by klonos** [comment]
A belated "yes!" to this one ...I almost got bitten by it again while on a content type deleting spree. When performing such a repetitive task, it is easy to miss the fact that a type has existing content, and instead treat it the same as the other content types (blindly hitting the "delete" button). I'll need to work on finalizing #3422 first though.
